arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,264 @@
+                 " Computing checksum for "
+              << tempPath;
+          auto checksumCalc = ChecksumCalculator();
+          if (TRI_ProcessFile(tempPath.c_str(),
+                              [&checksumCalc](char const* buffer, size_t n) {
+                                checksumCalc.updateEVPWithContent(buffer, n);
+                                return true;
+                              })) {
+            checksumCalc.computeFinalChecksum();
+            writeShaFile(tempPath, checksumCalc.getChecksum());
+          }
+        }
+      }
+    }
+  }
+}
+
+std::string ChecksumHelper::removeFromTable(std::string const& fileName) {
+  std::string checksum;
+  {
+    MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+    if (auto it = _fileNamesToHashes.find(TRI_Basename(fileName));
+        it != _fileNamesToHashes.end()) {
+      checksum = it->second;
+      _fileNamesToHashes.erase(it);
+    }
+  }
+  return checksum;
+}
+
+std::string ChecksumHelper::buildShaFileNameFromSst(
+    std::string const& fileName, std::string const& checksum) {
+  if (!fileName.empty() && !checksum.empty()) {
+    TRI_ASSERT(fileName.size() > 4);
+    std::string shaFileName = fileName.substr(0, fileName.size() - 4);
+    TRI_ASSERT(!isFileNameSst(shaFileName));
+    shaFileName += ".sha." + checksum + ".hash";
+    return shaFileName;
+  }
+  return {};
+}
+
+rocksdb::Status ChecksumWritableFile::Append(const rocksdb::Slice& data) {
+  _checksumCalc.updateIncrementalChecksum(data.data(), data.size());
+  return rocksdb::WritableFileWrapper::Append(data);
+}
+
+rocksdb::Status ChecksumWritableFile::Close() {
+  TRI_ASSERT(_helper != nullptr);
+  _checksumCalc.computeFinalChecksum();
+  if (!_helper->writeShaFile(_fileName, _checksumCalc.getChecksum())) {
+    return rocksdb::Status::Corruption("File writing was unsuccessful");
+  }
+  return rocksdb::WritableFileWrapper::Close();
+}
+
+rocksdb::Status ChecksumEnv::NewWritableFile(
+    const std::string& fileName, std::unique_ptr<rocksdb::WritableFile>* result,
+    const rocksdb::EnvOptions& options) {
+  std::unique_ptr<rocksdb::WritableFile> writableFile;
+  rocksdb::Status s =
+      rocksdb::EnvWrapper::NewWritableFile(fileName, &writableFile, options);
+  if (!s.ok()) {
+    return s;
+  }
+  try {
+    if (_helper->isFileNameSst(fileName)) {
+      result->reset(
+          new ChecksumWritableFile(std::move(writableFile), fileName, _helper));
+    } else {
+      *result = std::move(writableFile);
+    }
+  } catch (std::exception& e) {
+    LOG_TOPIC("8b19e", ERR, arangodb::Logger::ENGINES)
+        << "WritableFile: exception caught when allocating " << e.what();
+    return rocksdb::Status::MemoryLimit(
+        "WritableFile: exception caught when allocating");
+  }
+  return s;
+}
+
+rocksdb::Status ChecksumEnv::DeleteFile(const std::string& fileName) {
+  if (_helper->isFileNameSst(fileName) ||
+      fileName.find(".sha") != std::string::npos) {
```

Comment:
Will RocksDB ever call us to delete a `.sha` file? I guess not, because it is not aware of these files.

Reply:
I agree. And the code following this will also not work for `.sha` files, as far as I see. I think RocksDB is not aware of the `.sha` files and thus will not call us to delete any.

Reply:
Will remove sha from the if